consensus/hotstuff/timeoutaggregator/timeout_aggregator.go

```diff
@@ -0,0 +1,231 @@
+// No errors are expected during normal operations.
+func NewTimeoutAggregator(log zerolog.Logger,
+	notifier hotstuff.Consumer,
+	lowestRetainedView uint64,
+	committee hotstuff.Replicas,
+	collectors hotstuff.TimeoutCollectors,
+) (*TimeoutAggregator, error) {
+	queuedTimeouts, err := fifoqueue.NewFifoQueue(fifoqueue.WithCapacity(defaultTimeoutQueueCapacity))
+	if err != nil {
+		return nil, fmt.Errorf("could not initialize timeouts queue")
+	}
+
+	aggregator := &TimeoutAggregator{
+		log:                        log,
+		notifier:                   notifier,
+		lowestRetainedView:         counters.NewMonotonousCounter(lowestRetainedView),
+		activeView:                 counters.NewMonotonousCounter(lowestRetainedView),
+		committee:                  committee,
+		collectors:                 collectors,
+		queuedTimeoutsNotifier:     engine.NewNotifier(),
+		enteringViewEventsNotifier: engine.NewNotifier(),
+		queuedTimeouts:             queuedTimeouts,
+	}
+
+	componentBuilder := component.NewComponentManagerBuilder()
+	for i := 0; i < defaultTimeoutAggregatorWorkers; i++ { // manager for worker routines that process inbound events
+		componentBuilder.AddWorker(func(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
+			ready()
+			aggregator.queuedTimeoutsProcessingLoop(ctx)
+		})
+	}
+	componentBuilder.AddWorker(func(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
+		ready()
+		aggregator.enteringViewProcessingLoop(ctx)
+	})
+
+	aggregator.ComponentManager = componentBuilder.Build()
+	return aggregator, nil
+}
+
+// queuedTimeoutsProcessingLoop is the event loop which waits for notification about pending work
+// and as soon as there is some it triggers processing.
+// All errors are propagated to SignalerContext
+func (t *TimeoutAggregator) queuedTimeoutsProcessingLoop(ctx irrecoverable.SignalerContext) {
+	notifier := t.queuedTimeoutsNotifier.Channel()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-notifier:
+			err := t.processQueuedTimeoutEvents(ctx)
+			if err != nil {
+				ctx.Throw(fmt.Errorf("internal error processing queued timeout events: %w", err))
+				return
+			}
+		}
+	}
+}
+
+// processQueuedTimeoutEvents is a function which dispatches previously queued timeouts on worker thread
+// This function is called whenever we have queued timeouts ready to be dispatched.
+// No errors are expected during normal operations.
+func (t *TimeoutAggregator) processQueuedTimeoutEvents(ctx context.Context) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		default:
+		}
+
+		msg, ok := t.queuedTimeouts.Pop()
+		if ok {
+			timeoutObject := msg.(*model.TimeoutObject)
+			err := t.processQueuedTimeout(timeoutObject)
+			if err != nil {
+				return fmt.Errorf("could not process pending TO %v: %w", timeoutObject.ID(), err)
+			}
+
+			t.log.Info().
+				Uint64("view", timeoutObject.View).
+				Str("timeout_id", timeoutObject.ID().String()).
+				Msg("TO has been processed successfully")
```

Comment:
same here: TO's ID is probably of little interest but expensive to compute. On the other hand, we don't include the signer, which is probably relevant:
```suggestion
				Hex("signer", timeoutObject.SignerID[:]).
				Msg("TO has been processed successfully")
```